subsys/zigbee/osif/zb_nrf_transceiver.c

```diff
@@ -257,332 +311,234 @@ zb_bool_t zb_trans_transmit(zb_uint8_t wait_type, zb_time_t tx_at,
+enum net_verdict ieee802154_radio_handle_ack(struct net_if *iface,
+					     struct net_pkt *pkt)
 {
-#ifdef CONFIG_RADIO_STATISTICS
-	zb_osif_get_radio_stats()->rx_successful++;
-#endif /* defined CONFIG_RADIO_STATISTICS */
+	ARG_UNUSED(iface);
 
-	rx_frame_t *rx_frame_free_slot = NULL;
+	size_t ack_len = net_pkt_get_len(pkt);
 
-	/* Find free slot for frame */
-	for (uint32_t i = 0; i < ARRAY_SIZE(rx_frames); i++) {
-		if (rx_frames[i].data == NULL) {
-			rx_frame_free_slot = &rx_frames[i];
-			break;
-		}
+	if (ack_len != ACK_PKT_LENGTH) {
+		LOG_ERR("%s: ACK length error", __func__);
+		return NET_CONTINUE;
 	}
 
-	if (rx_frame_free_slot == NULL) {
-		__ASSERT(false,
-			 "Not enough rx frames allocated for 15.4 driver");
-		return;
+	if ((*net_pkt_data(pkt) & FRAME_TYPE_MASK) != FRAME_TYPE_ACK) {
+		LOG_ERR("%s: ACK frame was expected", __func__);
+		return NET_CONTINUE;
 	}
 
-	rx_frame_free_slot->data = data;
-	rx_frame_free_slot->power = power;
-	rx_frame_free_slot->lqi = lqi;
-	rx_frame_free_slot->time = time;
+	if (ack_frame != NULL) {
+		LOG_ERR("Overwriting unhandled ACK frame.");
+	}
 
-	if (data[ACK_REQUEST_OFFSET] & ACK_REQUEST_BIT) {
-		rx_frame_free_slot->pending_bit = acked_with_pending_bit;
-	} else {
-		rx_frame_free_slot->pending_bit = ZB_FALSE;
+	ack_frame_buf[0] = ack_len;
+	if (net_pkt_read(pkt, &ack_frame_buf[1], ack_len) < 0) {
+		LOG_ERR("Failed to read ACK frame.");
+		return NET_CONTINUE;
 	}
 
-	acked_with_pending_bit = ZB_FALSE;
-	k_fifo_put(&rx_fifo, rx_frame_free_slot);
+	/* ack_frame != NULL informs that ACK frame has been received */
+	ack_frame = ack_frame_buf;
+
+	return NET_OK;
+}
+
+static enum net_verdict zigbee_l2_recv(struct net_if *iface,
+					struct net_pkt *pkt)
+{
+	ARG_UNUSED(iface);
+
+	k_fifo_put(&rx_fifo, pkt);
 
 	zb_macll_set_rx_flag();
 	zb_macll_set_trans_int();
 
-	/* Raise signal to indicate radio event */
+	/* Raise signal to indicate rx event */
 	zigbee_event_notify(ZIGBEE_EVENT_RX_DONE);
-}
 
-/* Callback notify that reception of a frame failed. */
-void nrf_802154_receive_failed(nrf_802154_rx_error_t error)
-{
-	acked_with_pending_bit = ZB_FALSE;
-
-#ifdef CONFIG_RADIO_STATISTICS
-	switch (error) {
-	case NRF_802154_RX_ERROR_NONE:
-		zb_osif_get_radio_stats()->rx_err_none++;
-		break;
-	case NRF_802154_RX_ERROR_INVALID_FRAME:
-		zb_osif_get_radio_stats()->rx_err_invalid_frame++;
-		break;
-	case NRF_802154_RX_ERROR_INVALID_FCS:
-		zb_osif_get_radio_stats()->rx_err_invalid_fcs++;
-		break;
-	case NRF_802154_RX_ERROR_INVALID_DEST_ADDR:
-		zb_osif_get_radio_stats()->rx_err_invalid_dest_addr++;
-		break;
-	case NRF_802154_RX_ERROR_RUNTIME:
-		zb_osif_get_radio_stats()->rx_err_runtime++;
-		break;
-	case NRF_802154_RX_ERROR_TIMESLOT_ENDED:
-		zb_osif_get_radio_stats()->rx_err_timeslot_ended++;
-		break;
-	case NRF_802154_RX_ERROR_ABORTED:
-		zb_osif_get_radio_stats()->rx_err_aborted++;
-		break;
-	}
-#endif /* defined CONFIG_RADIO_STATISTICS */
+	return NET_OK;
```

Comment:
Do I properly understand that the `CONFIG_RADIO_STATISTICS` functionality will be dropped by this PR?

Reply:
We have some limited access to tx error after my PR https://github.com/nrfconnect/sdk-zephyr/pull/357. When it comes to rx errors, there's currently no way to access rx error code as even shim driver doesn't save it in radio driver callback.
Because of this I think collecting radio statistics has to be dropped.
I'll remove `CONFIG_RADIO_STATISTICS` from Kconfig if you don't have objections on this.

Reply:
removed